### PR TITLE
feat: Add lora_grpo implementation to training hub algorithms

### PR DESCRIPTION
**Pull request by Fiona-Waters**
[RHOAIENG-62928](https://redhat.atlassian.net/browse/RHOAIENG-62928) 

**What this PR does / why we need it**:

Add LORA_GRPO algorithm for RLVR/GRPO training
Adds support for Group Relative Policy Optimization (GRPO) via the ART backend as a new Training Hub algorithm.

**Changes**
Algorithm registration (algorithms.py)

 - Register lora_grpo in the algorithm registry with DEFAULT_COMMAND entrypoint (not torchrun — ART is single-GPU and manages its own vLLM subprocess; torchrun env vars leak into it and cause NCCL timeouts)
 - Metrics file pattern: training_metrics.jsonl (shared with lora_sft)
RHAI Training Hub integration (rhai/traininghub.py)

 - Add LORA_GRPO to TrainingHubAlgorithms enum
 - Add _read_lora_grpo_metrics() — reads ART's two-phase JSONL format (rollout + train lines per iteration), merges them into a single dict
 - Add _transform_lora_grpo() — maps GRPO metrics to controller schema (progressPercentage, currentStep, loss, mean_reward, full_match_rate, entropy)
 - Wrap instrumentation activation in if __name__ == '__main__': guard to prevent [Errno 98] Address already in use when ART's multiprocessing.spawn re-executes the script in child processes

**Tests**

 - Algorithm registry: spec lookup, known algorithms set, shared metrics pattern allowlist, pod metadata, python entrypoint assertion
 - RHAI: enum value, entrypoint validation, wrapper script __main__ guard, two-phase metrics reader, empty file handling, metrics transformer (with and without max_steps), cleanup on startup

**Dependencies**
 - Training Hub 0.8.1 release

**Which issue(s) this PR fixes** _(optional, in `Fixes #<issue number>, #<issue number>, ...` format, will close the issue(s) when PR gets merged)_:

Fixes #

**Verification steps** _(required; provide valid, reproducible steps for reviewers to verify your changes, e.g. setup, commands, expected outcome, notebooks)_:

_Example: 1. Check out this branch. 2. Run `make test-python`. 3. Confirm that …_

**Checklist:**

- [ ] [Docs](https://www.kubeflow.org/docs/) included if any changes are user facing

## Summary by CodeRabbit

* **New Features**
  * Added LoRA GRPO training algorithm with integrated metrics reading and merging, mapping to controller-friendly training metrics (includes progress percentage and remaining-time estimates).
  * Generated runtime wrappers now include a main-entry guard and use a Python entrypoint for improved runtime compatibility and stability.

* **Tests**
  * Extended test coverage for LoRA GRPO: metrics instrumentation, two-phase merged metrics handling, stale/empty metrics behavior, progress/remaining-time calculations, and runtime-wrapper behavior.

### CODE REVIEW

File: kubeflow/trainer/rhai/traininghub.py [hunk 1]
```diff
@@ -573,6 +577,91 @@ def _transform_lora_sft(self, metrics):
...
+                "currentStep": step,
+                "totalSteps": max_steps if max_steps else None,
+                "currentEpoch": max(1, math.ceil(epoch)),
+                "totalEpochs": None,
+                "trainMetrics": {
+                    "loss": f"{loss_val:.4f}" if loss_val is not None else None,
+                    "learning_rate": f"{lr_val:.6f}" if lr_val is not None else None,
+                    "grad_norm": f"{grad_norm_val:.4f}" if grad_norm_val is not None else None,
+                    "mean_reward": f"{mean_reward_val:.4f}"
+                    if mean_reward_val is not None
```
Comment (MStokluska) [inline kubeflow/trainer/rhai/traininghub.py]: nit: could formatting style be consistent here?

File: kubeflow/trainer/rhai/traininghub.py [hunk 2]
```diff
@@ -573,6 +577,91 @@ def _transform_lora_sft(self, metrics):
...
+            full_match_rate_val = metrics.get("full_match_rate")
+            entropy_val = metrics.get("entropy")
+
+            return {
+                "progressPercentage": percent_int,
+                "estimatedRemainingSeconds": None,
+                "currentStep": step,
+                "totalSteps": max_steps if max_steps else None,
+                "currentEpoch": max(1, math.ceil(epoch)),
+                "totalEpochs": None,
```
Comment (MStokluska) [inline kubeflow/trainer/rhai/traininghub.py]: Nit: If I understand correctly, totalEpochs is always 1 (it's iteration based algorithm, not epoch based). So maybe we could set total to 1 always?

File: kubeflow/trainer/rhai/traininghub.py [hunk 3]
```diff
@@ -573,6 +577,91 @@ def _transform_lora_sft(self, metrics):
...
+            loss_val = metrics.get("loss")
+            lr_val = metrics.get("learning_rate")
+            grad_norm_val = metrics.get("grad_norm")
+            mean_reward_val = metrics.get("mean_reward")
+            full_match_rate_val = metrics.get("full_match_rate")
+            entropy_val = metrics.get("entropy")
+
+            return {
+                "progressPercentage": percent_int,
+                "estimatedRemainingSeconds": None,
```
Comment (MStokluska) [inline kubeflow/trainer/rhai/traininghub.py]: Could we not use the `wall_time_s` for calculating the remaining seconds?

**Review by MStokluska — COMMENTED** [review]
Thanks Fiona!
I've run this on cluster and apart from known issue it works as expected.
Just couple minor points on metrics.

File: kubeflow/trainer/rhai/traininghub.py (hunk 1 above)
Comment (Fiona-Waters) [reply kubeflow/trainer/rhai/traininghub.py]: Thanks, updated

File: kubeflow/trainer/rhai/traininghub.py (hunk 2 above)
Comment (Fiona-Waters) [reply kubeflow/trainer/rhai/traininghub.py]: You're right, updated to 1

File: kubeflow/trainer/rhai/traininghub.py (hunk 3 above)
Comment (Fiona-Waters) [reply kubeflow/trainer/rhai/traininghub.py]: Yes we can, updated.

File: kubeflow/trainer/rhai/traininghub.py [hunk 4]
```diff
@@ -573,6 +577,100 @@ def _transform_lora_sft(self, metrics):
...
+                if not os.path.exists(metrics_file):
+                    return {}
+
+                try:
+                    result = subprocess.run(
+                        ["tail", "-n", "2", metrics_file],
+                        capture_output=True,
+                        text=True,
+                        check=True,
+                    )
```
[the remark this answers is not in the document]
Comment (Fiona-Waters) [reply kubeflow/trainer/rhai/traininghub.py:599]: Updated to fix.

File: kubeflow/trainer/rhai/traininghub.py [hunk 5]
```diff
@@ -573,6 +577,100 @@ def _transform_lora_sft(self, metrics):
...
+                merged = {}
+                for line in lines:
+                    line = line.strip()
+                    if line:
+                        try:
+                            merged.update(json.loads(line))
+                        except json.JSONDecodeError:
+                            continue
+
+                return merged
```
[the remark this answers is not in the document]
Comment (Fiona-Waters) [reply kubeflow/trainer/rhai/traininghub.py]: Updated to fix.

File: kubeflow/trainer/rhai/traininghub.py [hunk 6]
```diff
@@ -818,16 +916,19 @@ def training_func(func_args):
...
+        params_lines: list[str] = ["if __name__ == '__main__':\n    training_func({\n"]
         for key, value in func_args.items():
-            params_lines.append(f"    {repr(key)}: {repr(value)},\n")
-        params_lines.append("})\n")
+            params_lines.append(f"        {repr(key)}: {repr(value)},\n")
+        params_lines.append("    })\n")
         call_line = "".join(params_lines)
     else:
-        call_line = f"training_func({func_args})\n"
+        call_line = f"if __name__ == '__main__':\n    training_func({func_args})\n"
```
[the remark this answers is not in the document]
Comment (Fiona-Waters) [reply kubeflow/trainer/rhai/traininghub.py:942]: Updated to fix.

File: kubeflow/trainer/rhai/traininghub.py [hunk 7]
```diff
@@ -573,6 +577,110 @@ def _transform_lora_sft(self, metrics):
...
+            metrics_file = f"{ckpt_output_dir}/{metrics_file_rank0}"
+
+            try:
+                if not os.path.exists(metrics_file):
+                    return {}
+
+                try:
+                    result = subprocess.run(
+                        ["tail", "-n", "2", metrics_file],
+                        capture_output=True,
```
[the remark this answers is not in the document]
Comment (Fiona-Waters) [reply kubeflow/trainer/rhai/traininghub.py:595]: The existing metrics readers for SFT, OSFT, and LoRA SFT all use the same subprocess.run(["tail", ...]) pattern (lines 255, 314, 355). Keeping the GRPO reader consistent with the existing code for now. The risk is minimal since the code runs inside a training container that already executes user-provided training functions.

**Review by briangallagher — APPROVED** [review]
